src/resolver.ts

```diff
@@ -3130,12 +3138,62 @@ export class Resolver extends DiagnosticEmitter {
               }
             }
             if (!fieldType) break; // did report above
+            if (existingField !== null) {
+              // visibility checks
+              let baseClass = <Class>base;
+              let thisFieldIsPubOrProt = !fieldPrototype.is(CommonFlags.PRIVATE);
+              let existingFieldIsPublic = existingField.isPublic;
+
+              if (thisFieldIsPubOrProt && !existingFieldIsPublic) {
+                this.errorRelated(
+                  DiagnosticCode.Property_0_is_private_in_type_1_but_not_in_type_2,
+                  fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                  fieldPrototype.name, baseClass.internalName, instance.internalName
+                );
+              } else if (fieldPrototype.is(CommonFlags.PROTECTED) && existingFieldIsPublic) {
+                this.errorRelated(
+                  DiagnosticCode.Property_0_is_protected_in_type_1_but_public_in_type_2,
+                  fieldPrototype.identifierNode.range, existingField.identifierNode.range,
+                  fieldPrototype.name, instance.internalName, baseClass.internalName
+                );
+              } else if (!thisFieldIsPubOrProt && existingFieldIsPublic) {
```

Comment:
There seems to be a missing case where `private < protected`. I might be missing something, though, but in general I'd prefer if these checks were a little easier to grasp. Seems the rules here are:

* If existing field is private, it cannot be redeclared at all
* If existing field is protected, it can be redeclared as either protected or public
* If existing field is public, it can only be redeclared as public
* Plus: If existing field has an initializer but the redeclared one has not, that's an error

(that correct?)

Reply:
You're completely right, I'll make the checks a bit easier to grasp

> Plus: If existing field has an initializer but the redeclared one has not, that's an error

This doesn't throw an error:
```ts
class A {
  public a: number = 1;
  constructor(a: number) { this.a = 2; }
}

class B extends A {
  constructor(public a: number) { super(a) }
}
```
perhaps you're referring to if the redeclared field is redundant but has an initializer, no error is thrown?

Reply:
In particular I was wondering about the diagnostic

> Property_0_will_overwrite_the_base_property_in_1_If_this_is_intentional_add_an_initializer_Otherwise_remove_the_redundant_declaration

which seems to refer to the case

```ts
class A {
  public foo = 1;
}
class B extends A {
  public foo;
}
```

while if it is a ctor parameter as well then an initializer is kinda implied. (but I have actually no idea)

Reply:
The two cases can be told apart btw by looking at `Field.parameterIndex`. If it is `-1` it is just a field (as in my snippet), otherwise it is a field that is also a constructor parameter at the specified index.

Reply:
agreed to scrap the `Property_0_will_overwrite_the_base_property_in_1_If_this_is_intentional_add_an_initializer_Otherwise_remove_the_redundant_declaration` error